Sources/WeaverDI/Builder.swift

```diff
@@ -33,15 +51,31 @@ final class Builder {
         }
         
         return { (parameters: () -> P) -> I in
-            self.locker.lock()
-            defer { self.locker.unlock() }
-            
-            if let instance = self.instance?.value as? I {
+
+            if self.syncIsSet {
+                guard let instance = self.instance?.value as? I else {
+                    fatalError()
+                }
                 return instance
             }
+
+            self.instanceLocker.lock()
             
+            if self.syncIsSet {
+                guard let instance = self.instance?.value as? I else {
+                    fatalError()
+                }
+                self.instanceLocker.unlock()
+                return instance
+            }
+
             let instance = body(parameters)
             self.instance = Instance(value: instance as AnyObject, scope: self.scope)
+            
+            self.syncIsSet = true
```

Comment:
This is a race condition. `syncIsSet` is actually asynchronous, leading to `instanceLocker` getting unlocked before `isSet` is actually set.

The current thread needs to wait until `isSet` is set before unlocking `instanceLocker`